src/librustdoc/lib.rs

```diff
@@ -698,15 +704,10 @@ fn run_renderer<'tcx, T: formats::FormatRenderer<'tcx>>(
 ) -> MainResult {
     match formats::run_format::<T>(krate, renderopts, cache, tcx) {
         Ok(_) => tcx.sess.has_errors().map_or(Ok(()), Err),
-        Err(e) => {
-            let mut msg =
-                tcx.sess.struct_err(format!("couldn't generate documentation: {}", e.error));
-            let file = e.file.display().to_string();
-            if !file.is_empty() {
-                msg.note(format!("failed to create or modify \"{}\"", file));
-            }
-            Err(msg.emit())
-        }
+        Err(e) => Err(tcx.sess.emit_err(CouldntGenerateDocumentation {
+            error: e.error,
+            file: e.file.display().to_string(),
```

Comment:
Is `file` ever empty? I think it may be better to represent this as `Option<String>` so you can just annotate it with `#[note]`

Reply:
This is another alternative that I found, it requires adding another struct for a `Subdiagnostic` note.

```rust
#[derive(Diagnostic)]
#[diag(rustdoc_couldnt_generate_documentation)]
pub(crate) struct CouldntGenerateDocumentation {
    pub(crate) error: String,
    #[subdiagnostic]
    pub(crate) file: Option<FailedToCreateOrModifyFile>,
}

#[derive(Subdiagnostic)]
#[note(rustdoc_failed_to_create_or_modify_file)]
pub(crate) struct FailedToCreateOrModifyFile {
    pub(crate) file: String,
}
```

Reply:
Having `Option<String>` for the `file` field (in the version specified in this review comment) and adding `#[note]` to the struct causes compilation failures.

This is because `Option<String>` doesn't implement `IntoDiagnosticArgs`. Implementing `IntoDiagnosticArgs` for `Option<String>` is also not allowed due to the orphan rules. Creating a new (wrapped) type for `Option<String>`, and then implementing the trait wouldn't work because the `Option` is "gone" from the struct field.

CMIIW here :)